File: RemoteControlManager.m
//
//  RemoteControlManager.m
//  SmartDeviceLink-Example-ObjC
//
//  Created by Beharry, Justin (J.S.) on 8/1/22.
//  Copyright © 2022 smartdevicelink. All rights reserved.
//

#import "RemoteControlManager.h"
#import "SmartDeviceLink.h"
#import "AlertManager.h"

@interface RemoteControlManager()

@property (strong, nonatomic) SDLManager *sdlManager;
@property (strong, nonatomic) SDLRemoteControlCapabilities *remoteControlCapabilities;
@property (strong, nonatomic) NSString *climateModuleId;
@property (strong, nonatomic) NSNumber<SDLBool> *consent;

- (NSArray*)getButtons;

@end

@implementation RemoteControlManager

- (instancetype)initWithManager:(SDLManager *)manager {
    self = [super init];
    if (!self) {
        return nil;
    }

    _sdlManager = manager;
    
    [_sdlManager.systemCapabilityManager subscribeToCapabilityType:SDLSystemCapabilityTypeRemoteControl withUpdateHandler:^(SDLSystemCapability * _Nullable capability, BOOL subscribed, NSError * _Nullable error) {
        if (!capability.remoteControlCapability) { return; }
        
        NSLog(@"Retrieving Remote Control Capabilities");
        self->_remoteControlCapabilities = capability.remoteControlCapability;
        self->_climateModuleId = self->_remoteControlCapabilities.climateControlCapabilities.firstObject.moduleInfo.moduleId;
                
        /// Get consent to control modules
        SDLGetInteriorVehicleDataConsent *getInteriorVehicleDataConsent = [[SDLGetInteriorVehicleDataConsent alloc] initWithModuleType:SDLModuleTypeClimate moduleIds:@[self->_climateModuleId]];
        [self.sdlManager sendRequest:getInteriorVehicleDataConsent withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
            if (!response.success) { return; }
            
            NSLog(@"Retrieving Consent");
            NSLog(@"Bool Array %@", response.success);
            self->_consent = response.success;
        }];
    }];

    return self;
}

- (void)showClimateControl {
    if (_climateModuleId == NULL && _consent == 0) {
        NSString *errorMessage = @"The climate module id was not set or consent was not given";
        [AlertManager sendAlertWithManager:_sdlManager image:nil textField1:errorMessage textField2:nil];
    }
    
    [_sdlManager.screenManager beginUpdates];
    _sdlManager.screenManager.softButtonObjects = [self getButtons];
    [_sdlManager.screenManager endUpdates];
}

- (NSString*)getClimateData {
    NSString* helloWorld = _climateModuleId;
    return helloWorld;
}

- (NSArray*)getButtons {
    SDLSoftButtonObject *acOnButton = [[SDLSoftButtonObject alloc] initWithName:@"AC On" text:@"AC On" artwork:nil handler:^(SDLOnButtonPress * _Nullable buttonPress, SDLOnButtonEvent * _Nullable buttonEvent) {
        if (buttonPress == nil) { return; }
        NSLog(@"Turn AC On");
    }];
    
    SDLSoftButtonObject *acOffButton = [[SDLSoftButtonObject alloc] initWithName:@"AC Off" text:@"AC Off" artwork:nil handler:^(SDLOnButtonPress * _Nullable buttonPress, SDLOnButtonEvent * _Nullable buttonEvent) {
        if (buttonPress == nil) { return; }
        NSLog(@"Turn AC Off");
    }];
    
    SDLSoftButtonObject *acMaxToggle = [[SDLSoftButtonObject alloc] initWithName:@"AC Max" text:@"AC Max" artwork:nil handler:^(SDLOnButtonPress * _Nullable buttonPress, SDLOnButtonEvent * _Nullable buttonEvent) {
        if (buttonPress == nil) { return; }
        NSLog(@"Toggle AC Max");
    }];
    
    SDLSoftButtonObject *temperatureDecreaseButton = [[SDLSoftButtonObject alloc] initWithName:@"Temperature Decrease" text:@"Temperature -" artwork:nil handler:^(SDLOnButtonPress * _Nullable buttonPress, SDLOnButtonEvent * _Nullable buttonEvent) {
        if (buttonPress == nil) { return; }
        NSLog(@"Temperature -");
    }];
    
    SDLSoftButtonObject *temperatureIncreaseButton = [[SDLSoftButtonObject alloc] initWithName:@"Temperature Increase" text:@"Temperature +" artwork:nil handler:^(SDLOnButtonPress * _Nullable buttonPress, SDLOnButtonEvent * _Nullable buttonEvent) {
        if (buttonPress == nil) { return; }
        NSLog(@"Temperature +");
    }];
    
    SDLSoftButtonObject *setClimateButton = [[SDLSoftButtonObject alloc] initWithName:@"Set Climate" text:@"Set Climate" artwork:nil handler:^(SDLOnButtonPress * _Nullable buttonPress, SDLOnButtonEvent * _Nullable buttonEvent) {
        if (buttonPress == nil) { return; }
        NSLog(@"Set Climate");
    }];
    
    return @[acOnButton, acOffButton, acMaxToggle, temperatureDecreaseButton, temperatureIncreaseButton, setClimateButton];
}

@end

File: RemoteControlManager.swift
//
//  RemoteControlManager.swift
//  SmartDeviceLink-Example-Swift
//
//  Created by Beharry, Justin (J.S.) on 7/28/22.
//  Copyright © 2022 smartdevicelink. All rights reserved.
//

import Foundation
import SmartDeviceLink

class RemoteControlManager {
    private var sdlManager: SDLManager!
    private var remoteControlCapabilities: SDLRemoteControlCapabilities!
    private var climateModuleId: String!
    private var consent: [Bool]!
    
    
    init(sdlManager: SDLManager) {
        self.sdlManager = sdlManager
        /// Retrieve remote control information and store module ids
        self.sdlManager.systemCapabilityManager.subscribe(capabilityType: .remoteControl) { (capability, subscribed, error) in
            guard capability?.remoteControlCapability != nil else { return }
            self.remoteControlCapabilities = capability?.remoteControlCapability
            
            
            let firstClimateModule = self.remoteControlCapabilities.climateControlCapabilities?.first
            let moduleId = firstClimateModule?.moduleInfo?.moduleId
            
            self.climateModuleId = moduleId
            print("Climate Control Module Id ------------------------------------------------------------------------- \(self.climateModuleId ?? "Missing")")

            /// Get Consent to control module
            let getInteriorVehicleDataConsent = SDLGetInteriorVehicleDataConsent(moduleType: .climate, moduleIds: [self.climateModuleId])
            self.sdlManager.send(request: getInteriorVehicleDataConsent, responseHandler: { (request, response, error) in
                guard let res = response as? SDLGetInteriorVehicleDataConsentResponse else { return }
                guard let allowed = res.allowed else { return }
                let boolAllowed = allowed.map({ (bool) -> Bool in
                    return bool.boolValue
                })
                
                self.consent = boolAllowed
                
                self.showClimateControl()
            })
        }
    }
    
    private func subscribeVehicleData() {
        sdlManager.subscribe(to: .SDLDidReceiveInteriorVehicleData) { (message) in
            guard let onInteriorVehicleData = message as? SDLOnInteriorVehicleData else { return }
            
            print("Climate Data Changing")
            print(onInteriorVehicleData.moduleData.climateControlData as Any)
        }
    }
    
    private func subscribeClimateData() {
        let getInteriorVehicleData = SDLGetInteriorVehicleData(andSubscribeToModuleType: .climate, moduleId: climateModuleId)
        self.sdlManager.send(request: getInteriorVehicleData) { (req, res, err) in
            guard let response = res as? SDLGetInteriorVehicleDataResponse else { return }
            
            print(response)
        }
    }
    
    func turnOnAC() {
        let climateDictionary: [String: Any] = [
            "acEnable": true as NSNumber & SDLBool
        ]
                    
        let climateControlData = SDLClimateControlData(dictionary: climateDictionary)
        let moduleData = SDLModuleData(climateControlData: climateControlData)
        let setInteriorVehicleData = SDLSetInteriorVehicleData(moduleData: moduleData)

        self.sdlManager.send(request: setInteriorVehicleData) { (request, response, error) in
            guard response?.success.boolValue == true else { return }
        }
    }
    
    func turnOffAC() {
        let climateDictionary: [String: Any] = [
            "acEnable": false as NSNumber & SDLBool
        ]
                    
        let climateControlData = SDLClimateControlData(dictionary: climateDictionary)
        let moduleData = SDLModuleData(climateControlData: climateControlData)
        let setInteriorVehicleData = SDLSetInteriorVehicleData(moduleData: moduleData)

        self.sdlManager.send(request: setInteriorVehicleData) { (request, response, error) in
            guard response?.success.boolValue == true else { return }
        }
    }
    
    func showClimateControl() {
        let screenManager = self.sdlManager.screenManager
        screenManager.beginUpdates()
        
        screenManager.textField1 = "Climate Control Data"
        screenManager.textField2 = "Press Buttons"
        screenManager.textField3 = "Remote Control"
        
        self.subscribeVehicleData()
        self.subscribeClimateData()
        
        
        screenManager.changeLayout(SDLTemplateConfiguration(predefinedLayout: .nonMedia)) { err in
            // This listener will be ignored, and will use the handler set in the endUpdates call.
        }
        
        let acOnButton = SDLSoftButtonObject(name: "AC On", text: "AC On", artwork: nil) { (buttonPress, buttonEvent) in
            guard buttonPress != nil else { return }
            
            self.turnOnAC()
        }
        
        let acOffButton = SDLSoftButtonObject(name: "AC Off", text: "AC Off", artwork: nil) { (buttonPress, buttonEvent) in
            guard buttonPress != nil else { return }
                        
            self.turnOffAC()
        }
        
        screenManager.softButtonObjects = [acOnButton, acOffButton]
                
        screenManager.endUpdates { error in
            if error != nil {
                // Print out the error if there is one and return early
                print("Issue updating UI")
                return
            }
        }
    }
}
